// @shelf/jest-mongodb@6.0.0/lib/environment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const fs_1 = require("fs");
const crypto_1 = require("crypto");
const jest_environment_node_1 = require("jest-environment-node");
const mongodb_memory_server_1 = require("mongodb-memory-server");
const helpers_1 = require("./helpers");
// eslint-disable-next-line @typescript-eslint/no-require-imports
const debug = require('debug')('jest-mongodb:environment');
module.exports = class MongoEnvironment extends jest_environment_node_1.TestEnvironment {
    globalConfigPath;
    mongo;
    constructor(config, context) {
        super(config, context);
        this.globalConfigPath = (0, path_1.join)(config.globalConfig.rootDir, 'globalConfig.json');
        const options = (0, helpers_1.getMongodbMemoryOptions)(config.globalConfig.rootDir);
        const isReplSet = (0, helpers_1.isMongoMemoryReplSetOptions)(options);
        debug(`isReplSet`, isReplSet);
        this.mongo = isReplSet ? new mongodb_memory_server_1.MongoMemoryReplSet(options) : new mongodb_memory_server_1.MongoMemoryServer(options);
    }
    async setup() {
        debug('Setup MongoDB Test Environment');
        const globalConfig = JSON.parse((0, fs_1.readFileSync)(this.globalConfigPath, 'utf-8'));
        if (globalConfig.mongoUri) {
            this.global.__MONGO_URI__ = globalConfig.mongoUri;
        }
        else {
            await this.mongo.start();
            this.global.__MONGO_URI__ = this.mongo.getUri();
        }
        this.global.__MONGO_DB_NAME__ = globalConfig.mongoDBName || (0, crypto_1.randomUUID)();
        await super.setup();
    }
    async teardown() {
        debug('Teardown MongoDB Test Environment');
        await this.mongo.stop();
        await super.teardown();
    }
    // @ts-expect-error - runScript has incompatible type definitions
    runScript(script) {
        // @ts-expect-error - parent class method call
        return super.runScript(script);
    }
};
